prelude/src/stdlib/other/IO.sk

```diff
@@ -105,90 +145,113 @@ mutable class File(
 }
 
 mutable class StringIO(
-  mutable buffer: String = "",
-  mutable pos: Int = 0,
-) uses Readable, Writable, IsTerminal {
+  private buffer: mutable Vector<UInt8> = mutable Vector[],
+  private mutable pos: Int = 0,
+) uses Read, Write, IsTerminal {
   fun isatty(): Bool {
     false
   }
 
-  mutable fun read(sz: Int): String {
-    start = this.buffer.getIter().forward(this.pos);
-    if (sz == 0) {
-      return ""
+  mutable fun read(buf: mutable MutableBytes): Result<Int, Error> {
+    readSz = min(buf.size(), this.buffer.size() - this.pos);
+    // TODO: Proper copy.
+    for (i in Range(0, readSz)) {
+      buf![i] = this.buffer[this.pos + i]
     };
-    readSz = min(sz, this.buffer.length() - this.pos);
     this.!pos = this.pos + readSz;
-    res = this.buffer.sub(start, readSz);
-    if (res.length() == 0 && sz != 0) {
-      throw EndOfFile()
-    };
-    res
-  }
 
-  mutable fun read_to_end(): String {
-    start = this.buffer.getIter().forward(this.pos);
-    this.!pos = this.buffer.length();
-    this.buffer.substring(start)
+    Success(readSz)
   }
 
-  mutable fun write(s: String): Int {
-    this.!buffer = this.buffer + s;
-    s.length()
+  mutable fun write(buf: readonly Bytes): Result<Int, Error> {
+    this.buffer.extend(buf);
+    Success(buf.size())
   }
+}
 
-  mutable fun write_all(s: String): void {
-    this.!buffer = this.buffer + s;
+trait BufRead extends Read {
+  mutable fun fill_buf(): Result<readonly Bytes, Error>;
+  mutable fun consume(nbytes: Int): void;
+
+  mutable fun has_data_left(): Result<Bool, Error> {
+    this.fill_buf() match {
+    | Success(buf) -> Success(buf.size() > 0)
+    | Failure(err) -> Failure(err)
+    }
```

Comment:
I think that won't work because the type of `this.fill_buf()` is `Result<readonly Bytes, Error>` and we want to return a `Result<Bool, Error`, hence the somewhat cumbersome `Failure(err) -> Failure(err)`.